### DISCUSSION TITLE
Is there a way to make the picking faster?

### DISCUSSION THREAD

**Discussion by ssaornil** [visgl/deck.gl#5793] [Q&A]
I have four pickable layers (PathLayer and ScatterplotLayers) and it takes a very long time to obtain the picked object.

![picking](https://user-images.githubusercontent.com/18212144/118961710-e97bb980-b964-11eb-879a-ac30103434f7.png)

The first half is the picking performed by the `onHover` event. The second half is our code where we want to obtain all picked objecst from all layers using `pickMultipleObjects`.

Is there a way to make this picking faster? It's way too long now.

**Comment by Pessimistress** [comment]
Some ideas:

- If you're using a large `pickingRadius` it could lead to longer `readPixels` time
- You can block the default picking using [layerFilter](https://deck.gl/docs/api-reference/core/deck#layerfilter):

```js
layerFilter: ({renderPass}) => renderPass !== 'picking:hover'
```

- if you have to call `pickMultipleObjects` on hover, apply a debounce.

If any picking is generally taking a long time - how many objects do you have on the pickable layers?

**Reply by ssaornil** [reply]
Thanks!

The `layerFilter` option had no effect because we cannot filter out any of them. We kind of need them all to be pickable since they hold different information.
We will try reducing the radius, we use a 7 pixel radius at the moment. I am not sure adding a debounce will have any effect on a single pick object query. It will definitely help with having less queries but the problem of the time it takes to obtain a single item is still too long.

In terms of number of objects, we are talking about 16-20K on each layer which sometimes end up in 50K instances in the internal state for the PathLayer.

**Reply by Pessimistress** [reply]
The `layerFilter` I suggested only filters out layers during the default picking pass on hover. If I'm not mistaken, you are not using the picking result in the pickingInfo object passed to `onHover`, instead you are using the result from your own `pickMultipleObjects` call. Therefore you should set Deck's `pickingRadius` prop to 1 and apply the suggested `layerFilter` to minimize the impact of the default picking pass.

**Reply by ssaornil** [reply]
I missunderstood the aim of using `layerFilter`. With that approach we would only perform the picking once which will reduce the time in half, that's brilliant!. Thanks @Pessimistress.

What would you recommend instead of using the picking radius? Would it make sense to have an extra layer with all elements with a bigger size but not visible (maybe opacity 0 or transparent color) to be the pickable layer? I think I saw this approach in an example somewhere.

**Reply by Pessimistress** [reply]
Deck's pickingRadius prop only affects the default picking behavior. You should still set your custom picking radius when you call the picking method yourself.
